### DISCUSSION TITLE
使用UE5.2源碼版編譯項目development server報錯

### DISCUSSION THREAD

**Discussion by RAPUTADEV** [endink/Mediapipe4u-plugin#116] [General]
### MediaPipe4U 版本

20230811

### UnrealEngine 版本

5.2.x

### UnrealEngine 类型

源码编译

### 问题类型

项目打包

### 错误详细描述

將Mediapipe置於項目的Plugins中
Build Development Editor可以成功
但在Build Development Server時出現以下錯誤:

>Missing precompiled manifest for 'MediaPipe', 'C:\Users\Raputa\Dharma_source\Intermediate\Build\Win64\Dharmaverse\Debug\MediaPipe\MediaPipe.precompiled'. This module was most likely not flagged for being included in a precompiled build - set 'PrecompileForTargets = PrecompileTargetsType.Any;' in MediaPipe.build.cs to override. If part of a plugin, also check if its 'Type' is correct.

VisualStudio:
[builddevelopmentserver.txt](https://github.com/endink/Mediapipe4u-plugin/files/12342677/builddevelopmentserver.txt)
UnrealBuildTool Log:
[Log.txt](https://github.com/endink/Mediapipe4u-plugin/files/12342700/Log.txt)

另外測試將Mediapipe移除
項目Build Developmment Editor和 Development Server都可以成功

**Comment by endink** [comment]
是否阅读过源码引擎打包，预编译插件需要手动复制 C++ link 文件。   

https://opensource.labijie.com/Mediapipe4u-plugin/package/source_engine_package.html

**Comment by RAPUTADEV** [comment]
有照文檔複製過C++ link
development editor打包成功的
但是development server打包報錯

**Comment by endink** [comment]
能否让我知道你的打包 `development server` 如何复现？

**Comment by endink** [comment]
切换到 Dedicated Server ，玩家数量设置为 2 ？我需要一个你更改过的设置以复现你的问题（不怕辛苦可以新建一个项目，然后复现这个问题，然后告诉我如何复现）

**Comment by RAPUTADEV** [comment]
我將Mediapipe放入項目/plugins
執行copy_intermediate.bat 
Generate Visual Studio project file
打開項目.sln
在VisualStudio裡選擇Target為Development Editor/Win64 選擇到Game/項目 右鍵選擇build --> build成功 -->能打開項目 並且確認項目打開時Mediapipe plugins都是勾選狀態

接下來照著UE官方文檔，在項目/source中加入了[項目]Server.Target.cs 如下
```
using UnrealBuildTool;
using System.Collections.Generic;
using System;

[SupportedPlatforms(UnrealPlatformClass.Server)]
public class DharmaverseServerTarget : TargetRules
{
    public DharmaverseServerTarget(TargetInfo Target) : base(Target)
    {
        Type = TargetType.Server;
        DefaultBuildSettings = BuildSettingsVersion.V2;
        //IncludeOrderVersion = EngineIncludeOrderVersion.Latest;

        ExtraModuleNames.AddRange(new string[] { "Dharmaverse" });

        //DharmaverseTarget.ApplySharedLyraTargetSettings(this);

        //bUseChecksInShipping = true;
    }
}
```
接著重新Generate Visual Studio project file，打開項目.sln
在VisualStudio裡選擇Target為Development Server/Win64，選擇到項目 右鍵選擇build --> build失敗 出現錯誤訊息
```
Missing precompiled manifest for 'MediaPipe', 'C:\Users\Raputa\Dharma_source\Intermediate\Build\Win64\Dharmaverse\Debug\MediaPipe\MediaPipe.precompiled'. This module was most likely not flagged for being included in a precompiled build - set 'PrecompileForTargets = PrecompileTargetsType.Any;' in MediaPipe.build.cs to override. If part of a plugin, also check if its 'Type' is correct.
```

**Comment by endink** [comment]
感谢，稍后会用你的步骤复现，看你的打包日志可能需要手动复制连接符号来打包：

`[Project Plugins]`\MediaPipe4U\MediaPipe4U\Intermediate\Build\Win64\UnrealGame\Development\MediaPipe   
复制到   
C:\Users\Raputa\Dharma_source\Intermediate\Build\Win64\DharmaverseServer\Development\MediaPipe   

可以写一个脚本

**Comment by endink** [comment]
源码编译后的引擎，总是不能自己复制 C++ 符号文件，或许你可以尝试将插件放到`Engine\Plugins\Marketplace` ， 我不确定这样是否可以成功打包。

> 这个插件和普通插件的区别只是它是预编译的，不包含源代码，无法打包是因为源码编译的引擎中处理预编译插件打包时会丢掉一些动作，这些动作只有手动操作，或许你将插件放到 `Engine\Plugins\Marketplace` 可以完成自动打包
